### ISSUE TITLE
Panic on removing and re-adding a server to a cluster

### ISSUE THREAD

**Issue by ripienaar** [nats-io/nats-server#3494]
When a node is added to a cluster and resources are assigned to it sometimes a panic can happen. 

We have a healthy R3 cluster all running 74c0b18f with many assets, we wish to replace a server so:

 * Stop the server completely 
 * `nats server raft peer-remove n1-lon` to remove it from the raft group
 * rm -rf the data dir so there's no old state left
 * Start the server again using the `n1-lon` name as before.

It recovers, a number of streams get assigned to it and it syncs data, so far so good.  But then sometimes after a few tries at this I will see a panic like this which then is unrecoverable (keeps panicing on every boot)

```
Sep 23 13:27:47 n1-lon nats-server[1329]: [1] [INF] JetStream cluster new remote metadata leader
Sep 23 13:27:47 n2-lon nats-server[1336]: [1] [WRN] RAFT [ACbIFjhc - C-R3F-QHj1HjNu] Request from follower for index [130] possibly beyond our last index [139] - no message found
Sep 23 13:27:47 n1-lon nats-server[1329]: [1] [WRN] RAFT [4FFRXkOw - C-R3F-QHj1HjNu] Expected first catchup entry to be a snapshot and peerstate, will retry
Sep 23 13:27:47 n2-lon nats-server[1336]: [1] [WRN] RAFT [ACbIFjhc - C-R3F-ZNRU0x64] Request from follower for index [145] possibly beyond our last index [153] - no message found
Sep 23 13:27:47 n1-lon nats-server[1329]: [1] [WRN] RAFT [4FFRXkOw - C-R3F-ZNRU0x64] Expected first catchup entry to be a snapshot and peerstate, will retry
Sep 23 13:27:47 n1-lon nats-server[1329]: [1] [WRN] Error applying entries to 'one > SFO_MIRROR': last sequence mismatch
Sep 23 13:27:47 n1-lon nats-server[1329]: [1] [WRN] Resetting stream cluster state for 'one > SFO_MIRROR'
Sep 23 13:27:47 n1-lon nats-server[1329]: [1] [WRN] Error applying entries to 'one > ORDERS_27': last sequence mismatch
Sep 23 13:27:47 n1-lon nats-server[1329]: panic: assignment to entry in nil map
Sep 23 13:27:47 n1-lon nats-server[1329]: goroutine 200 [running]:
Sep 23 13:27:47 n1-lon nats-server[1329]: github.com/nats-io/nats-server/v2/server.(*stream).setConsumer(0xc00032f800, 0xc000ffed80)
Sep 23 13:27:47 n1-lon nats-server[1329]: #011/home/runner/work/nats-server/src/github.com/nats-io/nats-server/server/stream.go:4495 +0x45
Sep 23 13:27:47 n1-lon nats-server[1329]: github.com/nats-io/nats-server/v2/server.(*stream).addConsumerWithAssignment(0xc00032f800, 0xc001384ea0, {0xc000029976, 0x2}, 0xc0010e5ef0, 0x1?)
Sep 23 13:27:47 n1-lon nats-server[1329]: #011/home/runner/work/nats-server/src/github.com/nats-io/nats-server/server/consumer.go:817 +0x1b4e
Sep 23 13:27:47 n1-lon nats-server[1329]: github.com/nats-io/nats-server/v2/server.(*jetStream).processClusterCreateConsumer(0xc00030a000, 0xc0010e5ef0, 0x0, 0x1)
Sep 23 13:27:47 n1-lon nats-server[1329]: #011/home/runner/work/nats-server/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:3430 +0x4ce
Sep 23 13:27:47 n1-lon nats-server[1329]: github.com/nats-io/nats-server/v2/server.(*jetStream).processConsumerAssignment(0xc00030a000, 0xc0010e5ef0)
Sep 23 13:27:47 n1-lon nats-server[1329]: #011/home/runner/work/nats-server/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:3278 +0x8de
Sep 23 13:27:47 n1-lon nats-server[1329]: github.com/nats-io/nats-server/v2/server.(*jetStream).applyMetaEntries(0xc00030a000, {0xc0000a9258, 0x1, 0x0?}, 0x0)
Sep 23 13:27:47 n1-lon nats-server[1329]: #011/home/runner/work/nats-server/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:1473 +0x352
Sep 23 13:27:47 n1-lon nats-server[1329]: github.com/nats-io/nats-server/v2/server.(*jetStream).monitorCluster(0xc00030a000)
Sep 23 13:27:47 n1-lon nats-server[1329]: #011/home/runner/work/nats-server/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:952 +0xaa8
Sep 23 13:27:47 n1-lon nats-server[1329]: created by github.com/nats-io/nats-server/v2/server.(*Server).startGoRoutine
Sep 23 13:27:47 n1-lon nats-server[1329]: #011/home/runner/work/nats-server/src/github.com/nats-io/nats-server/server/server.go:3077 +0x85
Sep 23 13:27:47 n2-lon nats-server[1336]: [1] [INF] 46.101.56.171:50174 - rid:1100 - Router connection closed: Read Error
Sep 23 13:27:47 n3-lon nats-server[1347]: [1] [ERR] 46.101.56.171:36010 - rid:1161 - Error flushing: writev tcp 172.17.0.2:6222->46.101.56.171:36010: writev: connection reset by peer
Sep 23 13:27:47 n3-lon nats-server[1347]: [1] [INF] 46.101.56.171:36010 - rid:1161 - Router connection closed: Write Error
Sep 23 13:27:49 n3-lon nats-server[1347]: [1] [ERR] Error trying to connect to route (attempt 2): dial tcp 46.101.56.171:6222: i/o timeout
Sep 23 13:27:49 n2-lon nats-server[1336]: [1] [ERR] Error trying to connect to route (attempt 2): dial tcp 46.101.56.171:6222: i/o timeout
```

It's different streams on different tries that trigger it

**Comment by ripienaar** [comment]
I left it to panic crash looping and after about 7 crashes it stopped and was helathy

**Comment by ripienaar** [comment]
I also noticed the cluster in question just never recovered really it was consuming a lot CPU after this

<img width="509" alt="Screenshot 2022-09-23 at 16 00 46" src="https://user-images.githubusercontent.com/82342/191978102-0ad2389d-612d-489f-a902-9471e8cf73bf.png">


I restarted the cluster completely and got a lot (many pages) of this for multiple streams.

```
Sep 23 13:59:52 n3-lon nats-server[1347]: [1] [INF] JetStream cluster new stream leader for 'one > ORDERS_9'
Sep 23 13:59:52 n3-lon nats-server[1347]: [1] [WRN] RAFT [J0t0ySUw - S-R3F-v5LcJgBY] Request from follower for index [1002769] possibly beyond our last index [1002771] - no message found
Sep 23 13:59:52 n1-lon nats-server[1329]: [1] [WRN] RAFT [4FFRXkOw - S-R3F-v5LcJgBY] Expected first catchup entry to be a snapshot and peerstate, will retry
Sep 23 13:59:53 n1-nyc nats-server[1333]: [1] [INF] JetStream cluster new consumer leader for 'one > ORDERS_70 > C2'
Sep 23 13:59:53 n1-nyc nats-server[1333]: [1] [WRN] RAFT [chC4jxAp - C-R3F-6dRLU6V2] Request from follower for index [131] possibly beyond our last index [132] - no message found
Sep 23 13:59:53 n3-nyc nats-server[1334]: [1] [INF] JetStream cluster new consumer leader for 'one > ORDERS_10 > C3'
Sep 23 13:59:53 n3-nyc nats-server[1334]: [1] [WRN] RAFT [P2IE71bg - C-R3F-e5Ri7XQG] Request from follower for index [131] possibly beyond our last index [132] - no message found
Sep 23 13:59:55 n3-sfo nats-server[1336]: [1] [INF] JetStream cluster new stream leader for 'one > ORDERS_38'
```

**Comment by kozlovic** [comment]
The panic is fixed by #3498. It is likely that the add/remove of the stream was replayed while the monitorConsumer go routine was running. Part of the remove would stop the stream, which in turn sets this map to nil when signaling the consumer that it should exit.
